libmamba/src/core/transaction.cpp

```diff
@@ -881,152 +876,86 @@ namespace mamba
-                        ));
-                        repr.separator.set_value("/");
-
-                        std::string total_str;
-                        if (dl_bar->total() == std::numeric_limits<std::size_t>::max())
-                        {
-                            total_str = "??.?MB";
-                        }
-                        else
-                        {
-                            total_str = to_human_readable_filesize(double(dl_bar->total()), 1);
-                        }
-                        repr.total.set_value(fmt::format("{:>7}", total_str));
-
-                        auto speed = dl_bar->avg_speed(std::chrono::milliseconds(500));
-                        repr.speed.set_value(
-                            speed
-                                ? fmt::format("@ {:>7}/s", to_human_readable_filesize(double(speed), 1))
-                                : ""
-                        );
-                    }
-                );
-            }
-
-            auto* extract_bar = aggregated_pbar_manager.aggregated_bar("Extract");
-            if (extract_bar)
-            {
-                extract_bar->set_repr_hook(
-                    [=](ProgressBarRepr& repr) -> void
-                    {
-                        auto active_tasks = extract_bar->active_tasks().size();
-                        if (active_tasks == 0)
-                        {
-                            repr.prefix.set_value(fmt::format("{:<16}", "Extracting"));
-                            repr.postfix.set_value(fmt::format("{:<25}", ""));
-                        }
-                        else
-                        {
-                            repr.prefix.set_value(fmt::format(
-                                "{:<11} {:>4}",
-                                "Extracting",
-                                fmt::format("({})", active_tasks)
-                            ));
-                            repr.postfix.set_value(
-                                fmt::format("{:<25}", extract_bar->last_active_task())
-                            );
-                        }
-                        repr.current.set_value(fmt::format("{:>3}", extract_bar->current()));
-                        repr.separator.set_value("/");
+            /*std::packaged_task<void(std::size_t)> task{[eit](std::size_t downloaded_size) {
+                eit->run(downloaded_size);
+            }};*/
+            auto task = std::make_shared<std::packaged_task<void(std::size_t)>>([eit](std::size_t downloaded_size) {
+                eit->run(downloaded_size);
+            });
+            extract_tasks.push_back(task->get_future());
+            download_requests.push_back(fit->build_download_request([extract_task=std::move(task)](std::size_t downloaded_size) {
+                MainExecutor::instance().schedule([t = std::move(extract_task)](std::size_t ds) {
+                    (*t)(ds); }, downloaded_size);
```

Comment:
@Klaim I find this part really cumbersome, `std::packaged_task` and `std::function` do not play well together. Having the `future` and the `promise` in the PackageExtractRequest would simplify this code a lot (although I admit it would expose synchronization primitives in the header :s).

Reply:
Yeah that's verbose because it's callback-based. Chaining continuations would be a bit less verbose but still noisy. The ultimate syntax to express the same thing would be through coroutines or task-graphs.

>  `std::packaged_task` and `std::function` do not play well together

I dont get what you mean here? `packaged_task` is just a callable object, it works well with `std::function`. Or do you mean the callback-based continuations is not ideal?

Reply:
`std::function` can store copiable objects only, while `std::package_task` (and therefore the lambda capturing it) is movable only. So we have to store the task in a `smart_pointer` and I really don't like it.

Reply:
Ah yes I'm used to have `std::move_only_function` around. In this case, dont use a smaprt pointer, just replaced `std::packaged_task` by a lambda wrapping a `std::promise`. Then if you need to, put the promise in the smart pointer. That avoids complicating the callback handling system.

```
auto p = std::make_shared<std::promise<void>>();
std::future<void> f = p->get_future();
auto continuation = [p = std::move(p), callback]{
   auto maybe_error = safe_invoke(callback);
   ... // error handling
   p->set_value();
 };
 ```
 
 Either that or we need an equivalent to `std::move_only_function` , there are many relyable implementations on github.

Reply:
let's go with the promise/future for now

Reply:
BTW if it's too much work, storing the `packaged_task` in a `shared_ptr` to pass it around should work too.

Reply:
Actually the version with promise instead of future does not simplify that much; we have 2 lambda instead of 3? but need to capture two elements and set manually the promise. Besides, we will get back to packaged_task when we introduce move_only_functions, so let's keep the packaged_task.

Reply:
You mean shared packated task? Works for me 👍🏽